fabric-component-access-api-v1/src/main/java/net/fabricmc/fabric/api/component/access/v1/ComponentType.java

```diff
@@ -0,0 +1,286 @@
+	 * by application of the given mapping function. Use this version for blocks that
+	 * may provide a component without the presence of a {@code BlockEntity}.
+	 *
+	 * <p>The mapping function should return {@link #absent()} if no component is available.
+	 *
+	 * @param mapping function that derives a component instance from an access context
+	 * @param blocks one or more blocks for which the function will apply
+	 */
+	void registerProvider(Function<BlockComponentContext, T> mapping, Block... blocks);
+
+	/**
+	 * Causes the given blocks to provide component instances of this type
+	 * via block entities associated with the given blocks.
+	 *
+	 * <p>Use this version for blocks where the {@code BlockEntity} *is* the component instance.
+	 *
+	 * @param blocks one or more blocks that will provide components in this way
+	 */
+	@SuppressWarnings("unchecked")
+	default void addProvider(Block... blocks) {
+		registerProvider(ctx -> (T) ctx.blockEntity(), blocks);
+	}
+
+	/**
+	 * Causes the given entity types to provide component instances of this type
+	 * by application of the given mapping function.
+	 *
+	 * <p>This will override any previous mapping of the same component type and only one
+	 * result per entity is possible.  For the reason, mod authors are advised to create
+	 * distinct component types for their use cases as needed to prevent conflicts.
+	 *
+	 * @param mapping mapping function that derives a component instance from an access context
+	 * @param entities one or more entities for which the function will apply
+	 */
+	void registerProvider(Function<EntityComponentContext, T> mapping, EntityType<?>... entities);
+
+	/**
+	 * Same as {@link #registerProvider(Function, EntityType...)} but matches all entities that
+	 * match the given predicate instead of providing specific entity types.  Use this to register
+	 * a provider for all entities that implement {@code LivingEntity}, for example.
+	 *
+	 * <p>This method may be called at any point during mod initialization (and not after) but the
+	 * predicate will only be applied after all registration is complete. Mod registration
+	 * order does not matter.
+	 *
+	 * @param mapping mapping function that derives a component instance from an access context
+	 * @param predicate Mapping will apply to all entity types that match this test
+	 */
+	void registerProvider(Function<EntityComponentContext, T> mapping, Predicate<EntityType<?>> predicate);
+
+	/**
+	 * Causes the given items to provide component instances of this type
+	 * by application of the given mapping function to item stacks having the given items.
+	 *
+	 * <p>The mapping function should return {@link #absent()} if no component is available.
+	 *
+	 * @param mapping mapping function that derives a component instance from an access context
+	 * @param items one or more items for which the function will apply
+	 */
+	void registerProvider(Function<ItemComponentContext, T> mapping, Item... items);
```

Comment:
I think we should clarify what each of these do with a specific method name, i.e. `registerItemProvider`, `registerBlockEntityProvider`

Reply:
done